src/app/AuthenticationController.tsx

```diff
@@ -193,10 +196,19 @@ class AuthenticationController extends React.Component<AuthenticationControllerP
         history.replace(this.props.landingRoute);
         this.props.setLandingRoute(undefined);
       }
-      this.setState({ stage: LoginStage.LOGGED_IN });
+      // Setting postLoginError to 'undefined' here ensures that when the user logs out
+      // after logging in successfully, they do not see the previous postLoginError message.
+      this.setState({ stage: LoginStage.LOGGED_IN, postLoginError: undefined });
     } catch (err) {
       console.error('Error on post-login actions.', err);
-      this.setState({ isPostLoginError: true });
+      this.setState({ postLoginError: AuthenticationController.PostLoginErrorMsg });
+      // Logging the user out immediately since they will need to log in again
+      // in order to retry the 'post-login' actions. The post-login actions
+      // must succeed before the user can proceed.
+      //
+      // When logout happens a re-render happens so need to track if the user has been previously
+      // logged out from this error message and display the message then.
+      this.props.logout();
```

Comment:
Also, you need to check if `logout()` succeeded and cannot call `this.setState({ postLoginError: AuthenticationController.PostLoginErrorMsg });` inconditionally.

For OpenShift, the back-end tries to invalidate tokens. See: https://github.com/kiali/kiali/blob/497614024c3a787ce4537121c03ae59cfb9f3e85/handlers/authentication.go#L743-L748 ...... Although I see session cookies are deleted even if OpenShift logout fails :-/ That's a probable bug in back-end's Logout function, because the UI won't transition to "logged out" state if the back-end throws an error (see [here](https://github.com/kiali/kiali-ui/blob/8472ae9bffc67f697342f368f1a722f2d52f323c/src/actions/LoginThunkActions.ts#L71-L79)).

But well.... ignoring that back-end bug... I wonder what state should the UI arrive on if logout fails? do we really want to blindly discard the error and assume proper logout (which is what you are doing now)?

Reply:
> But well.... ignoring that back-end bug... I wonder what state should the UI arrive on if logout fails? do we really want to blindly discard the error and assume proper logout (which is what you are doing now)?

This is a really good point and something I just overlooked so thanks for bringing it up. If logout fails the UI is now in a weird state because now the user is stuck logged in but without the ability to retry the "post-login" actions by logging in again. The user would need to refresh their browser in order to retry the "post-login" actions hence the original message telling the user to "refresh their session despite being logged in". But that was confusing since the actual problem was that one of the "post-login" requests was failing and "refreshing the session" wouldn't fix the problem if, for example, Prometheus was unreachable.

Taking a step back a bit, should these "post-login" requests really prevent the user from being able to log in to Kiali? If so, maybe they should be part of the log in process and log in should not succeed if these requests fail. If not, then Kiali does not need to prevent the user from proceeding after successful log in if any of the "post-login" requests fail. Kiali would load and the user would see some errors in the notifications area if any of the requests failed. 

My preference is to make the "post-login" requests non-blocking for successful login. That would get around the weird state of being logged in successfully but unable to log out and unable to proceed until the "post-login" requests succeed. Seems like a better UX and makes it more obvious what is going on since login succeeds but some services are down/unreachable so there are error notifications in the notifications area. @israel-hdez what do you think? Does Kiali _need_ the data from the "post-login" requests in order to load at all?

Reply:
> My preference is to make the "post-login" requests non-blocking for successful login.

After thinking about it and reviewing current code, I think the problem is that the fetched configs are stored in a "global" variable defined in this file: https://github.com/kiali/kiali-ui/blob/master/src/config/ServerConfig.ts.

So, we could make it non-blocking, but it's not that easy.

So, I understand that if we show the main console, some screens and controls may need a re-render after configs are fetched and don't match the default values. Since those configs are not stored in redux, it's hard to properly trigger a re-render. Also, some warning messages about missing config may need a delay - for example, the message telling that Jaeger is not configured. :-/

It's not a trivial change IMO.

> Does Kiali _need_ the data from the "post-login" requests in order to load at all?

Nope, we have some defaults, just in case.
It's more what I've told that the UI assumes valid configs once it boots the main console.

Reply:
> It's more what I've told that the UI assumes valid configs once it boots the main console.

I was afraid of that. What about showing the "Loading" screen while waiting for post login instead of booting to the main console? The user would remain logged in and eventually would see an error related to the post-login requests on the loading screen. They'd still be stuck on the loading screen until the post-login requests finish but imo it's more obvious at that point that the issue is not a "login" issue since they logged in successfully.

Reply:
That's an option. I guess rathern than going from login screen to main console, you would do two transitions, from login to loading to main console...

Although, I'm not sure if other will be OK with this transition. I remember this was the flow in the past and we removed the loading screen on first login to minimize screen flickering :-/

Reply:
> That's an option. I guess rathern than going from login screen to main console, you would do two transitions, from login to loading to main console...

Exactly

> Although, I'm not sure if others will be OK with this transition. I remember this was the flow in the past and we removed the loading screen on first login to minimize screen flickering :-/

Perhaps there can be some minimum time the loading screen appears? That's more complexity to handle in the loading screen but arguably less than getting into a state where you are logged in on the the login page but cannot log out and cannot proceed. I think having the loading screen more accurately expresses what the problem is to the user as well: "you've logged in successfully but Kiali can't initialize/load for some reason".

I'd still prefer the user being able to go straight from login --> console but given that will require more significant changes maybe that can be a gradual effort and the flow can be login --> loading --> console in the short term. Maybe this effort is more trouble than it's worth but I think it'll be harder to and handle all these various login stages at the login page and effectively communicate to the user what the problem is and how they can fix it.

Reply:
> Perhaps there can be some minimum time the loading screen appears?

If you ask to me, I'd say no.... I'm not too in favor of artificial delays :-D

I wonder if it's too hard to transition to the init screen only if there is an error? And if not, keep the current flow of login to main console?

or, perhaps, stay at login screen and replace the login button with a "refresh" button?

As you see, I'm not a UX expert. So, I'm just throwing suggestions to see if there is any better choice.

Reply:
> I wonder if it's too hard to transition to the init screen only if there is an error? And if not, keep the current flow of login to main console?

What about a timer on the post-login requests and when they are taking greater than some threshold (3 seconds?) then transition to the init screen? Waiting until the request failed could leave users on the login screen for a potentially long period of time in which the "login" button is stuck disabled. I like the transition to the init screen because it signals that login succeeded.

Perhaps just adding a ~5 second request timeout in the UI for the post-login actions along with a message telling them to refresh their page to retry once they have investigated and fixed the underlying issue would be good enough for now? Not performing any logout.

Reply:
> What about a timer on the post-login requests and when they are taking greater than some threshold (3 seconds?) then transition to the init screen?

I'm :+1: with this.

> Perhaps just adding a ~5 second request timeout in the UI for the post-login actions

I'm OK with this. Although, I'm not very sure, because such small timeout doesn't match the time the server can wait for external services.
But it's up to you. I don't see it wrong in any case.

Reply:
Added a timer that will transition to the loading page from login after 3 seconds during post-login.